Upgrade phpunit & fix tests

antoine-pous opened the pull request:
- [x] Update to PHPUnit 6
- - [x] RouteTest
- - [x] ConfigTest
- - [x] RouterTest
- - [x] composer.json
- [x] Travis-CI
- - [x] Add PHP 7.1
- - [x] Remove PHP version below 7.0
- [x] Validate tests
- - [x] ConfigTest 
- - [x] RouteTest
- - [x] RouterTest
- [x] Config optimizations
- - [x] Improved YAML format
- - [x] Implement JSON support

Commit:
Fix missing var action

Commit:
Fix wrong regexp

Commit:
Fix missing symfony/yaml dependency

Commit:
Upgrade test for PHPUnit 6

Commit:
Upgrade to PHPUnit 6 and test route without action

Commit:
Add PHP 7.1

Commit:
Updgrade to PHPUnit 6

Commit:
Allow underscore in filtername

Commit:
Fix getAction issue and implement getController method

Commit:
Add units tests for getController method

Commit:
Fix filtername typo

antoine-pous commented:
I need your advice on `testParamsWithoutDynamicFilterMatch`. Currently the test fail on 
```PHP
self::assertEquals(
    array(array('filename' => 'someJsFile.min')),
    $router->match('/js/someJsFile.min.js')->getParameters()
);

self::assertEquals(
    array(array('filename' => 'someJsFile.min.js')),
    $router->match('/js/someJsFile.min.js.js')->getParameters()
);
```

This is due to unsupported multiple matches.
`someJsFile` and `someJsFile.min` are matched and we have multiple choices about this behavior:

1. thrown an exception and force developpers to be more strict
1. always return the last match
1. support this kind of behavior

Commit:
Add phpunit/php-code-coverage dependency

malukenho commented on composer.json:
```diff
@@ -21,11 +21,13 @@
   "require-dev": {
     "malukenho/docheader": "^0.1.5",
-    "phpunit/phpunit": "4.8.*",
+    "phpunit/phpunit": "6.*",
```
use `^`

antoine-pous replied:
Fixed

malukenho commented on src/Route.php:
```diff
@@ -169,8 +181,8 @@ private function substituteFilter($matches)
-            if(!preg_match('~^:([[a-z]])$~i', $key)) {
-                throw new Exception('Invalid filter name `'.$key.'` it should contains only letters and start with `:`');
+            if(!preg_match('~^:([a-z_]+)$~i', $key)) {
+                throw new Exception('Invalid filter name `'.$key.'` it should contains only letters, underscore and start with `:`');
```
We can throw an specific exception here?

antoine-pous replied:
Wich exception would you throw exactly?

malukenho replied:
`InvalidArgumentException` or `RuntimeException` probably.

malukenho commented on src/Route.php:
```diff
@@ -187,24 +199,34 @@ public function setParameters(array $parameters)
-            call_user_func_array(array($instance, $this->action), $this->parameters);
+        if($action === '__construct') {
+            $instance = new $controller($this->parameters);
         } else {
```
can we remove the else?

malukenho commented on src/Route.php:
```diff
@@ -187,24 +199,34 @@ public function setParameters(array $parameters)
-        if (!is_null($this->action)) {
-            $instance = new $action[0];
-            call_user_func_array(array($instance, $this->action), $this->parameters);
+        if($action === '__construct') {
```
code style

malukenho commented on src/Route.php:
```diff
@@ -187,24 +199,34 @@ public function setParameters(array $parameters)
-            $instance = new $action[0];
-            call_user_func_array(array($instance, $this->action), $this->parameters);
+        if($action === '__construct') {
+            $instance = new $controller($this->parameters);
```
`class_exists`?

antoine-pous replied:
Why we should test if the class exists? This is note the job of PHPRouter...

malukenho replied:
It's not the job of a router instantiate a class and call it a method too... If that's the point just use a service container or some sort.

malukenho commented on src/Route.php:
```diff
@@ -187,24 +199,34 @@ public function setParameters(array $parameters)
         } else {
-            $instance = new $action[0]($this->parameters);
+            $instance = new $controller;
+            call_user_func_array(array($instance, $action), $this->parameters);
```
shall we use `[]`? Also, Why not drop support to php 5.*?

antoine-pous replied:
See #105

malukenho commented on src/Route.php:
```diff
@@ -187,24 +199,34 @@ public function setParameters(array $parameters)
 
     public function getAction()
     {
+        $action = explode('::', $this->config['_controller']);
```
move `::` to a constant.

malukenho commented on src/Route.php:
```diff
@@ -187,24 +199,34 @@ public function setParameters(array $parameters)
     {
+        $action = explode('::', $this->config['_controller']);
+
+        $this->controller = $action[0];
```
We probably should abstract or use any other package to instantiate objects for us. Look at the doctrine one.

antoine-pous replied:
I've added ne methods, with them PHPRouter is more flexible

malukenho commented on src/Route.php:
```diff
@@ -187,24 +199,34 @@ public function setParameters(array $parameters)
+
+    public function getController()
+    {
+        $this->getAction();
```
ouch!

antoine-pous replied:
Fixed

malukenho commented:
As for your question, it should stop on the first assertion and execute controller/method form there.

Commit:
Typo fix

Commit:
Move class::action separator to constant

Commit:
Improve editor config

Commit:
Fix indent and add scripts

Commit:
Fix coding style

antoine-pous commented:
@malukenho @dannyvankooten i don't understand why `$router->match()` use this:

```php
$currentDir = dirname($_SERVER['SCRIPT_NAME']);
//...
if ('/' !== $currentDir) {
    $requestUrl = str_replace($currentDir, '', $requestUrl);
}
```

When we use filter regex for JS file like `([[:alnum:].]+).js` on route `/js/:filename` the match is broken. This is because `dirname($_SERVER['SCRIPT_NAME']);` return `.`.

I really don't find any usage where this replacement is necessary, can you give me an example please?

Commit:
Update to phpunit 6

Commit:
Ignore build directory

Commit:
- Drop the php support below version 5.6

- Implement setFiltersRegex & getFiltersRegex + tests
- Implement setController & getController + tests
- Bugfixes

Commit:
Drop the php support below version 5.6

Commit:
Test route with wrong filtername

Commit:
Remove debug dumps

Commit:
Fix coding style errors

Commit:
Fix lasts tests

Commit:
Drop php support below version 5.6

Commit:
Rewrite file according standards

Commit:
Remove type hint, php5.6 incompatibility

antoine-pous commented:
All tests seem OK.

@malukenho Docheader dependency broke the last travis tests, see https://github.com/malukenho/docheader/issues/17 for more informations.

malukenho commented:
@antoine-pous As far as I can see your tests is failing due multiples issues.

First, `docheader` doesn't work with php versions bellow `5.5`. So that explains failures for version `5.3` and `5.4`.

Other versions are failing as unit tests is broken, that indicates that bugs were introduced. 
If you're having trouble with `docheader` on your machine. Check if dependencies are installed.

malukenho commented on .travis.yml:
```diff
@@ -7,28 +7,32 @@ cache:
-  - 5.3
-  - 5.4
-  - 5.5
   - 5.6
```
Can we drop support to 5.6?

antoine-pous replied:
PHP 5.6 is under security support until 31 Dec 2018

malukenho commented on .travis.yml:
```diff
@@ -7,28 +7,32 @@ cache:
-  - 5.4
-  - 5.5
   - 5.6
   - 7.0
```
I'd also drop support to `7.0`

antoine-pous replied:
PHP 7.0 is under active support until 3 Dec 2017 and security support until 3 Dec 2018

malukenho commented on composer.json:
```diff
@@ -1,36 +1,43 @@
-    "psr-4": {
-      "PHPRouter\\": "src/"
+    "require": {
+        "php": "^5.6|^7.0|^7.1",
```
just `^7.1`

antoine-pous replied:
See previous answers

malukenho commented on composer.json:
```diff
@@ -1,36 +1,43 @@
+        {
+            "name": "Jefersson Nathan",
+            "email": "malukenho@phpse.net"
+        }
```
Put your name is here.

malukenho commented on composer.json:
```diff
@@ -1,36 +1,43 @@
+    },
+    "require-dev": {
+        "malukenho/docheader": "^0.1.5",
+        "phpunit/phpunit": "~5.0",
```
We can use phpunit `^6.0` 😄

antoine-pous replied:
Not currently, due to the support of PHP5.6, but we can creat a branch for each major version of PHP no? :)

malukenho commented on src/Route.php:
```diff
@@ -55,13 +73,25 @@ class Route
+     * Regex used to validate filtername
+     * @var string
+     */
+    private $filtersRegex = ':([a-zA-Z_]+)';
```
use a constant.

malukenho commented on tests/src/PHPRouterTest/RouterTest.php:
```diff
@@ -158,6 +179,23 @@ public function testGenerateNotExistent()
+        $collection = new RouteCollection();
+        $route = new Route(
+            '/user/:user_id',
+            array(
```
`[]`

malukenho commented on tests/src/PHPRouterTest/RouterTest.php:
```diff
@@ -134,6 +134,27 @@ public function testParamsWithDynamicFilterMatch()
+        $collection = new RouteCollection();
+        $route = new Route(
+            '/test/params',
+            array(
```
`[]`

malukenho commented on phpunit.xml.dist:
```diff
@@ -1,35 +1,29 @@
-         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/4.8/phpunit.xsd"
-         backupGlobals="true"
+<phpunit bootstrap="tests/Bootstrap.php"
```
remove bootstrap, just use `vendor/autoload.php`

antoine-pous replied:
The tests crash with this value

malukenho replied:
You need to register the namespace using `autoload-dev`

malukenho commented on phpunit.xml.dist:
```diff
@@ -1,35 +1,29 @@
-         xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/4.8/phpunit.xsd"
-         backupGlobals="true"
+<phpunit bootstrap="tests/Bootstrap.php"
+         backupGlobals="false"
```
It's `false` by default right?

antoine-pous replied:
PHPUnit author made regulary changes, i want avoid problems due to configuration.

malukenho replied:
Just lock on a specifically version

malukenho commented on phpunit.xml.dist:
```diff
@@ -1,35 +1,29 @@
          convertWarningsToExceptions="true"
-         bootstrap="tests/Bootstrap.php">
-
+         processIsolation="false"
```
Isn't it `false` by default?

malukenho commented on phpunit.xml.dist:
```diff
@@ -1,35 +1,29 @@
-         bootstrap="tests/Bootstrap.php">
-
+         processIsolation="false"
+         stopOnFailure="false">
```
Isn't it `false` by default?

malukenho commented on phpunit.xml.dist:
```diff
@@ -1,35 +1,29 @@
-        <testsuite name="PHP Router">
-            <directory>tests</directory>
+        <testsuite name="PHPRouter Test Suite">
+            <directory>tests/src/PHPRouterTest/</directory>
```
leave it as `tests/` only

malukenho commented on phpunit.xml.dist:
```diff
@@ -1,35 +1,29 @@
     <logging>
-        <log type="coverage-text" target="php://stdout" showUncoveredFiles="true"/>
+        <log type="tap" target="build/report.tap"/>
+        <log type="junit" target="build/report.junit.xml"/>
```
`junit`?

antoine-pous replied:
oops my bad, i took phpunit config from my another project wich contain some JAVA :man_facepalming:

malukenho commented on phpunit.xml.dist:
```diff
@@ -1,35 +1,29 @@
-
     <logging>
-        <log type="coverage-text" target="php://stdout" showUncoveredFiles="true"/>
+        <log type="tap" target="build/report.tap"/>
```
are we using logs for some reason?

antoine-pous replied:
Personally i use them. I find the log easier to read and interpret than the terminal.

malukenho replied:
If it's not part of the `ci` we can remove it. also, you can use a local `phpunit.xml` with your own configs. Just leave what's relevant to the project here.

malukenho commented on src/Route.php:
```diff
@@ -18,7 +18,7 @@
 
 use Fig\Http\Message\RequestMethodInterface;
-use Exception;
+use InvalidArgumentException;
```
Should be on our own namespace.

malukenho commented on src/Route.php:
```diff
@@ -29,15 +29,33 @@ class Route
+     * Separator used between class and action
+     * @const string
+     */
+    const ACTION_SEPARATOR = '::';
```
`Paamayim Nekudotayim`

antoine-pous replied:
Oh! You're right :confused:

malukenho commented on src/Route.php:
```diff
@@ -29,15 +29,33 @@ class Route
+    ];
+
+    /**
+     * Controller class
```
remove this kind of comments that doesn't bring any clarification. instead rename the property to `$controllerClassName` or similar

antoine-pous replied:
Fixed

malukenho commented on src/Route.php:
```diff
@@ -29,15 +29,33 @@ class Route
+    private $controller;
+
+    /**
+    * Method called into the controller class
```
same as above

malukenho commented on src/Route.php:
```diff
@@ -81,12 +111,16 @@ class Route
-        $this->parameters = isset($config['parameters']) ? $config['parameters'] : array();
+        $this->url          = $resource;
+        $this->config       = $config;
+        $this->methods      = isset($config['methods']) ? (array) $config['methods'] : [];
```
use `??`

malukenho commented on src/Route.php:
```diff
@@ -81,12 +111,16 @@ class Route
+        $this->url          = $resource;
+        $this->config       = $config;
+        $this->methods      = isset($config['methods']) ? (array) $config['methods'] : [];
+        $this->target       = isset($config['target']) ? $config['target'] : null;
```
use `??`

malukenho commented on src/Route.php:
```diff
@@ -81,12 +111,16 @@ class Route
+        $this->config       = $config;
+        $this->methods      = isset($config['methods']) ? (array) $config['methods'] : [];
+        $this->target       = isset($config['target']) ? $config['target'] : null;
+        $this->name         = isset($config['name']) ? $config['name'] : null;
```
use `??`

malukenho commented on src/Route.php:
```diff
@@ -81,12 +111,16 @@ class Route
+        $this->methods      = isset($config['methods']) ? (array) $config['methods'] : [];
+        $this->target       = isset($config['target']) ? $config['target'] : null;
+        $this->name         = isset($config['name']) ? $config['name'] : null;
+        $this->customParams = isset($config['parameters']) ? (array) $config['parameters'] : [];
```
use `??`

malukenho commented on src/Route.php:
```diff
@@ -81,12 +111,16 @@ class Route
+
+        $action = explode(self::ACTION_SEPARATOR, $this->config['_controller']);
+        $this->controller = $action[0];
+        $this->action = !empty($action[1]) && trim($action[1]) !== '' ? $action[1] : '__construct';
```
what if class doesn't have a `__construct` method defined?
It'll crash badly. 

https://3v4l.org/qTm0F

malukenho commented on src/Route.php:
```diff
@@ -182,29 +216,98 @@ public function getParameters()
+     * @param string $action Action to call
+     * @return null
+     */
+    public function setAction($action)
```
We should avoid setters. please remove them.

antoine-pous replied:
Why?

malukenho commented on src/Route.php:
```diff
@@ -182,29 +216,98 @@ public function getParameters()
+     * @since 1.3.0
+     * @return string The regex used to match filters names
+     */
+    public function getFiltersRegex()
```
Nope

malukenho commented on src/Route.php:
```diff
@@ -182,29 +216,98 @@ public function getParameters()
+     * @param string The new regex used to match filters names
+     * @return null
+     */
+    public function setFiltersRegex($regex)
```
nope

antoine-pous replied:
I think we should let the developper decide wich is the regex used to validate filternames. Sometime it's usefull to use the same syntaxe from another major component of our project.

If it broke the router this is not our problem, we just give a method to set custom filtername validation.

malukenho commented on src/Router.php:
```diff
@@ -109,25 +109,24 @@ public function match($requestUrl, $requestMethod = RequestMethodInterface::METH
             }
 
-            if ('/' !== $currentDir) {
+            if (!in_array($currentDir, ['.', '/'])) {
```
missing third parameter

antoine-pous replied:
The third parameter is optionnal. Should we really fill in all the optional parameters?

malukenho replied:
Yep. Be strict at types.

malukenho commented on src/Router.php:
```diff
@@ -109,25 +109,24 @@ public function match($requestUrl, $requestMethod = RequestMethodInterface::METH
 
             $route = rtrim($routes->getRegex(), '/');
-            $pattern = '@^' . preg_quote($this->basePath) . $route . '/?$@i';
+            $pattern = '@^' . $this->basePath . $route . '/?$@i';
```
Missing `preg_quote`

malukenho commented on src/Router.php:
```diff
@@ -109,25 +109,24 @@ public function match($requestUrl, $requestMethod = RequestMethodInterface::METH
+            $params = [];
 
-            if (preg_match_all('/:([\w-%]+)/', $routes->getUrl(), $argument_keys)) {
+            if (preg_match_all('@' . $routes->getFiltersRegex() . '@', $routes->getUrl(), $argument_keys)) {
```
the wrapper is very inconsistency across the code

malukenho commented on tests/src/PHPRouterTest/RouteTest.php:
```diff
@@ -37,6 +38,18 @@ protected function setUp()
+            '/page/:page_id',
+            array(
+                '_controller' => 'PHPRouter\Test\SomeController',
+                'methods' => array(
```
`[]`

malukenho commented on tests/src/PHPRouterTest/RouteTest.php:
```diff
@@ -37,6 +38,18 @@ protected function setUp()
+        $this->routeWithoutAction = new Route(
+            '/page/:page_id',
+            array(
+                '_controller' => 'PHPRouter\Test\SomeController',
```
`::class`

malukenho commented on tests/src/PHPRouterTest/RouteTest.php:
```diff
@@ -92,5 +105,35 @@ public function testSetName()
     public function testGetAction()
     {
         self::assertEquals('page', $this->routeWithParameters->getAction());
+        self::assertEquals('__construct', $this->routeWithoutAction->getAction());
```
nope

malukenho commented on tests/src/PHPRouterTest/RouterTest.php:
```diff
@@ -158,6 +179,23 @@ public function testGenerateNotExistent()
+            )
+        );
+        $route->setFiltersRegex(':([a-z]+):');
+        $route->setFilters(array(':filename' => '([[:alnum:].]+).js'), true);
```
`[]`

malukenho commented on tests/src/PHPRouterTest/RouterTest.php:
```diff
@@ -108,13 +108,13 @@ public function testParamsWithDynamicFilterMatch()
-            '/js/:filename.js',
+            '/js/:filename',
             array(
                 '_controller' => 'PHPRouter\Test\SomeController::dynamicFilterUrlMatch',
```
`::class`

malukenho commented on tests/src/PHPRouterTest/RouterTest.php:
```diff
@@ -94,7 +94,7 @@ public function testGetParamsInsideControllerMethod()
             )
         );
-        $route->setFilters(array('page_id' => '([a-zA-Z]+)'), true);
+        $route->setFilters(array(':page_id' => '([a-zA-Z]+)'), true);
```
`[]`

malukenho commented on tests/src/PHPRouterTest/ConfigTest.php:
```diff
@@ -18,23 +18,21 @@
-            'InvalidArgumentException',
-            'The file fileNotExisting not exists!'
-        );
+        $this->expectException('InvalidArgumentException');
```
`::class`

malukenho commented on tests/src/PHPRouterTest/ConfigTest.php:
```diff
@@ -18,23 +18,21 @@
  * @package PHPRouter\Test
  */
-class ConfigTest extends PHPUnit_Framework_TestCase
+class ConfigTest extends TestCase
```
`final`

malukenho commented on src/Route.php:
```diff
@@ -182,29 +216,98 @@ public function getParameters()
 
     public function setParameters(array $parameters)
```
can we remove it?

antoine-pous replied:
Router use it :/

malukenho commented on .travis.yml:
```diff
@@ -7,28 +7,32 @@ cache:
 script:
-    - vendor/bin/phpunit -c phpunit.xml.dist
-    - vendor/bin/docheader check src/ tests/
+  - vendor/bin/phpunit -c phpunit.xml.dist
```
use the call defined on `composer.json`

malukenho commented on .travis.yml:
```diff
@@ -7,28 +7,32 @@ cache:
-    - vendor/bin/phpunit -c phpunit.xml.dist
-    - vendor/bin/docheader check src/ tests/
+  - vendor/bin/phpunit -c phpunit.xml.dist
+  - vendor/bin/docheader check src/ tests/
```
use the call defined on `composer.json`

malukenho commented on .travis.yml:
```diff
@@ -7,28 +7,32 @@ cache:
-    - vendor/bin/docheader check src/ tests/
+  - vendor/bin/phpunit -c phpunit.xml.dist
+  - vendor/bin/docheader check src/ tests/
+  - vendor/bin/phpcs --standard=psr2 src/
```
use the call defined on `composer.json`

malukenho commented on .travis.yml:
```diff
@@ -7,28 +7,32 @@ cache:
+  - vendor/bin/phpunit -c phpunit.xml.dist
+  - vendor/bin/docheader check src/ tests/
+  - vendor/bin/phpcs --standard=psr2 src/
+  - vendor/bin/phpunit --coverage-text --coverage-clover=coverage.clover
```
use the call defined on `composer.json`

malukenho commented on composer.json:
```diff
@@ -1,36 +1,43 @@
-      "name": "Danny van Kooten",
-      "email": "dannyvankooten@gmail.com",
-      "homepage": "http://dannyvankooten.com/"
+    "name": "dannyvankooten/php-router",
```
Why `retab` this file?

malukenho commented:
@antoine-pous if we are bring new things to this code, and refactor this code as you wish, may consider follow `psr-7`  and `psr-11`.

Commit:
Remove junit

Commit:
Fix tests directory

Commit:
Remove array oldstyle

Commit:
Add author

Commit:
Update tests namespace

Commit:
Officially drop PHP 5 support

Commit:
Implement JSON config file support

Commit:
Support filters & improve YAML format

Commit:
Implement JSON config file support

Commit:
Improve YAML format and implement JSON support

Commit:
Improve YAML and add JSON support - PHP7 required

Commit:
Drop PHP5 support

Commit:
Remove unwanted tests, PHPRouter\Route do the stuff

Commit:
Remove tap

Commit:
Typo fix

Commit:
Enable strict_types, add return type

Commit:
Improve typehints & return types
